### ISSUE TITLE
linuxdeployqt deploys dependencies of excluded/blacklisted libraries

### ISSUE THREAD

**Issue by zaps166** [probonopd/linuxdeployqt#390]
linuxdeployqt (42e51ea) deploys dependencies (recursively) of excluded/blacklisted libraries.

Example 1: `libEGL.so.1` is blacklisted, so it is not copied, but its dependency `libgbm.so.1` is copied (which breaks EGL on Mesa drivers on another computer).

Example 2: I want to exclude PulseAudio libraries from my application (because their absence doesn't break my app): `-exclude-libs=libpulse-simple.so.0,libpulse.so.0`. These libs are not copied, but their dependencies like `libpulsecommon-8.0.so`, `libsndfile.so.1` and then e.g. `libFLAC.so.8` are copied, but they shouldn't (orphans).

**Comment by probonopd** [comment]
Interesting. Are you sure the dependencies are not introduced by other libraries that are not blacklisted?

Example 1: Is there anything in the resulting AppDir that depends on `libgbm.so.1`? (`grep -r libgbm ./AppDir`)

Example 2: Is there anything in the resulting AppDir that depends on `libpulsecommon-8.0.so`, `libsndfile.so.1`, `libFLAC.so.8`?

Can you share a test AppImage and the script that created it?

**Comment by probonopd** [comment]
Does https://github.com/linuxdeploy/linuxdeploy-plugin-qt/ perform better in this regard?

**Comment by TheAssassin** [comment]
https://github.com/linuxdeploy/linuxdeploy/issues/83

**Comment by zaps166** [comment]
Ok, it is impossible to filter libs using `ldd`. I used `objdump -p` to check direct dependencies manually. `lddtree` (mentioned here: linuxdeploy/linuxdeploy#83) seems to be a good replacement.

---

I think `libgbm.so.1` should be blacklisted, it is part of Mesa.

**Comment by ilya-fedin** [comment]
Are there any progress on this? Having to specify recursive dependencies manually is really disappointing.

**Comment by probonopd** [comment]
Hello @ilya-fedin can you give a concrete example where this is hitting you?

**Comment by ilya-fedin** [comment]
That's required if one wants to have a good integration to the system. In other words, it's necessary to exclude gtk in order to use system theme (or user may get broken theming), libwayland (or the app will crash because system gtk is compiled with system libwayland) and etc, depending of what one wants to bundle.

**Comment by ilya-fedin** [comment]
I know there are #236, but it's not really a way, xcb extensions must be bundled.

**Comment by probonopd** [comment]
So @ilya-fedin if I understand you correctly, then what you are really saying is that you doubt https://github.com/AppImage/pkg2appimage/edit/master/excludelist is correct/optimal as it is and you would like to see another set of libraries to be excluded from being bundled. Happy to discuss that. Please open tickets in https://github.com/AppImage/pkg2appimage regarding changes you would like to see.

**Comment by ilya-fedin** [comment]
I don't think that would fit into excludelist in pkg2appimage. If one want to package a gtk application as appimage than he probably would expect that gtk is bundled. That is not the case for Qt that uses gtk just for theme integration.

**Comment by probonopd** [comment]
True. Then maybe I did not understand the problem yet. Which libraries do you want to exclude and why are they currently being bundled (linuxdeployqt only bundles libraries it detects with ldd).

**Comment by ilya-fedin** [comment]
Not to say that excludelist would be a too bureaucratic procedure when you bundiling custom Qt plugins for further integration. They would need excluding even more libraries that others would want to be bundled. It's really specific to app being bundled and there should be a reliable way to exclude libraries packager wants.

**Comment by ilya-fedin** [comment]
> Which libraries do you want to exclude

Since linuxdeployqt doesn't handle recursive libraries, I even can't do a reliable comparsion of what libraries I need to exclude to get result that I want.

**Comment by probonopd** [comment]
@ilya-fedin I understand that there is an issue and I would like to help but I fail what _exactly_ would need to be done to resolve the issue.

Would being able to extend the built-in exludelist manually address your needs? How could we prevent people from producing AppImages that run nowhere but on the developers' systems? (E.g., the developer may think that all systems have Wayland libraries installed whereas my systems don't.)

**Comment by ilya-fedin** [comment]
> Would being able to extend the built-in exludelist manually address your needs?

But `-exclude-libs` does exactly that (maybe not technically, but its goal seems to be excluding libraries as well), doesn't it?

> How could we prevent people from producing AppImages that run nowhere but on the developers' systems?

Oh. That's really hard, basically you can't prevent all errors, that's why fine tuning at packager level is needed, automatic machine can't know all the cases. If an app uses libwayland and libgtk for integration, then you can't bundle libgtk or theming will be broken if libgtk version is not compatible with system themes. But if you don't bundle libgtk, then you can't bundle libwayland and the app should be able (Qt is able) to fallback to XWayland if the library is not found. That would be still better then a crash due to a mix of libraries or broken theming.

**Comment by probonopd** [comment]
Agree, it's a [mess](https://media.ccc.de/v/ASG2018-174-2018_desktop_linux_platform_issues) ;-)

Actually, `-exclude-libs` seems to be doing just that, append additional entries to the built-in excludelist. (Unless we have a bug.)

https://github.com/probonopd/linuxdeployqt/blob/dbad0d305b94b78034ff4fb313b6fff922fb452f/tools/linuxdeployqt/shared.cpp#L447-L448

**Comment by ilya-fedin** [comment]
Maybe, even standard excludelist doesn't do recursive exclusion?
